lib/sync/datasetClientsCleaner.js

```diff
@@ -22,26 +23,42 @@ var storage;
 function DatasetClientsCleaner(opts) {
   this.retentionPeriod = opts.retentionPeriod || '24h';
   this.checkFrequency = opts.checkFrequency || '1h';
+  this.cleanerLockName = opts.cleanerLockName || 'locks:sync:DatasetCleaner';
+  this.lockTimeout = parseDuration(this.checkFrequency) / 2;
   this.job;
 }
 
-function cleanDatasetClients(retentionPeriod) {
-  var parsedTime = parseDuration(retentionPeriod);
-  async.waterfall([
-    async.apply(storage.listDatasetClients, {active: false}),
-    function filterDatasetClients(datasetClients, callback) {
-      var datasetClientsToRemove = _.filter(datasetClients, function(datasetClientJson) {
-        var datasetClient = DatasetClient.fromJSON(datasetClientJson);
-        return datasetClient.shouldBeRemoved(parsedTime);
-      });
-      return callback(null, datasetClientsToRemove);
-    },
-    async.apply(storage.removeDatasetClients)
-  ], function(err, deletedDatasetClients) {
+function cleanDatasetClients(retentionPeriod, lockName, lockTimeout) {
+  syncLock.acquire(lockName, lockTimeout, function(err, lockCode){
     if (err) {
-      debugError('Failed to cleanup dataset clients due to error %j', err);
+      debugError('Failed to acquire lock for key = %s : error = %s', lockName, err);
+    } else if(lockCode) {
+      debug('lock acquired for datasetClientCleaner code = %s. Start cleaning', lockCode);
+      var parsedTime = parseDuration(retentionPeriod);
+      async.waterfall([
+        async.apply(storage.listDatasetClients, {active: false}),
+        function filterDatasetClients(datasetClients, callback) {
+          var datasetClientsToRemove = _.filter(datasetClients, function(datasetClientJson) {
+            var datasetClient = DatasetClient.fromJSON(datasetClientJson);
+            return datasetClient.shouldBeRemoved(parsedTime);
+          });
+          return callback(null, datasetClientsToRemove);
+        },
+        async.apply(storage.removeDatasetClients)
+      ], function(err, deletedDatasetClients) {
+        syncLock.release(lockName, lockCode, function(unlockErr){
+          if (unlockErr) {
+            debugError('Failed to release lock due to error %s', unlockErr);
+          }
+        });
+        if (err) {
+          debugError('Failed to cleanup dataset clients due to error %j', err);
```

Comment:
About the point you highlighted in a previous PR. Should this be %s as %j doesn't seem to include the error message?